taichi/program/kernel.cpp

```diff
@@ -145,10 +142,97 @@ void Kernel::set_arg_int(int i, int64 d) {
   }
 }
 
+// XXX: sync with snode.cpp: fetch_reader_result
+static uint64 fetch_result_uint64(int i)
+{
+  uint64 ret;
+  auto arch = get_current_program().config.arch;
+  if (arch == Arch::cuda) {
+    // TODO: refactor
+    // XXX: what about unified memory?
+#if defined(TI_WITH_CUDA)
+    CUDADriver::get_instance().memcpy_device_to_host(&ret,
+        (uint64 *)get_current_program().result_buffer + i,
+        sizeof(uint64));
+#else
+    TI_NOT_IMPLEMENTED;
+#endif
+  } else if (arch_is_cpu(arch)) {
+    ret = ((uint64 *)get_current_program().result_buffer)[i];
+  } else {
+    ret = get_current_program().context.get_arg_as_uint64(i);
+  }
+  return ret;
+}
+
+template <typename T>
+static T fetch_result(int i) // TODO: move to Program::fetch_result
+{
+  return taichi_union_cast_with_different_sizes<T>(fetch_result_uint64(i));
+}
+
+float64 Kernel::get_ret_float(int i) {
+  auto dt = rets[i].dt;
+  if (dt == DataType::f32) {
+    return (float64)fetch_result<float32>(i);
+  } else if (dt == DataType::f64) {
+    return (float64)fetch_result<float64>(i);
+  } else if (dt == DataType::i32) {
+    return (float64)fetch_result<int32>(i);
+  } else if (dt == DataType::i64) {
+    return (float64)fetch_result<int64>(i);
+  } else if (dt == DataType::i8) {
+    return (float64)fetch_result<int8>(i);
```

Comment:
Would it be better to `TI_ERROR` when people are trying to read an integer return value as float? They can use `get_ret_int` instead.

Reply:
It's okay since these are ABIs and only used by our python code, an end-user never call this.

Reply:
Sounds good. I trust your decision.